File: backend/app/schemas/project.py
from datetime import datetime, date
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.project import ProjectStatus
from app.schemas.task import TaskResponse


class ProjectBase(BaseModel):
    """Base schema for project data."""
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = None
    client_name: str | None = Field(None, max_length=100)
    start_date: date | None = None
    due_date: date | None = None


class ProjectCreate(ProjectBase):
    """Schema for creating a project."""
    pass


class ProjectUpdate(BaseModel):
    """Schema for updating a project."""
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = None
    client_name: str | None = Field(None, max_length=100)
    status: ProjectStatus | None = None
    start_date: date | None = None
    due_date: date | None = None


class ProjectResponse(BaseModel):
    """Schema for project response."""
    id: UUID
    name: str
    description: str | None
    client_name: str | None
    status: ProjectStatus
    start_date: date | None
    due_date: date | None
    created_by: UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ProjectDetailResponse(ProjectResponse):
    """Schema for project detail response with tasks."""
    tasks: list[TaskResponse] = []


class ProjectListResponse(BaseModel):
    """Schema for paginated project list."""
    items: list[ProjectResponse]
    total: int
    page: int
    limit: int

File: frontend/src/types/project.ts
import { Task } from "./task";

export type ProjectStatus = "active" | "completed" | "archived";

export interface Project {
  id: string;
  name: string;
  description: string | null;
  client_name: string | null;
  status: ProjectStatus;
  start_date: string | null;
  due_date: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface ProjectDetail extends Project {
  tasks: Task[];
}

export interface ProjectListResponse {
  items: Project[];
  total: number;
  page: number;
  limit: number;
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
  client_name?: string;
  start_date?: string;
  due_date?: string;
}

export interface UpdateProjectRequest {
  name?: string;
  description?: string;
  client_name?: string;
  status?: ProjectStatus;
  start_date?: string;
  due_date?: string;
}
